toolchain/check/pattern_match.cpp

```diff
@@ -379,6 +380,69 @@ auto MatchContext::EmitPatternMatch(Context& context,
       }
       break;
     }
+    case CARBON_KIND(SemIR::TuplePattern tuple_pattern): {
+      auto subpattern_ids =
+          context.inst_blocks().Get(tuple_pattern.elements_id);
+      auto add_all_subscrutinees =
+          [&](llvm::ArrayRef<SemIR::InstId> subscrutinee_ids) {
+            for (auto [subpattern_id, subscrutinee_id] :
+                 llvm::reverse(llvm::zip(subpattern_ids, subscrutinee_ids))) {
+              AddWork({.pattern_id = subpattern_id,
+                       .scrutinee_id = subscrutinee_id});
+            }
+          };
+      if (!entry.scrutinee_id.has_value()) {
+        CARBON_CHECK(kind_ == MatchKind::Callee);
+        context.TODO(pattern.loc_id,
+                     "Support patterns besides bindings in parameter list");
+        break;
+      }
+      auto scrutinee = context.insts().GetWithLocId(entry.scrutinee_id);
+      if (auto scrutinee_literal =
+              scrutinee.inst.TryAs<SemIR::TupleLiteral>()) {
+        auto subscrutinee_ids =
+            context.inst_blocks().Get(scrutinee_literal->elements_id);
+        if (subscrutinee_ids.size() != subpattern_ids.size()) {
+          CARBON_DIAGNOSTIC(TuplePatternSizeDoesntMatchLiteral, Error,
+                            "tuple pattern expects {0} element{0:s}, but tuple "
+                            "literal has {1}",
+                            IntAsSelect, IntAsSelect);
+          context.emitter()
+              .Build(pattern.loc_id, TuplePatternSizeDoesntMatchLiteral,
+                     subpattern_ids.size(), subscrutinee_ids.size())
+              .Emit();
```

Comment:
```suggestion
              .Emit(pattern.loc_id, TuplePatternSizeDoesntMatchLiteral,
                     subpattern_ids.size(), subscrutinee_ids.size());
```

The separate build call is only needed when adding notes.